### PR TITLE
Graceful handling of ReadTimeoutExceptions

### PR DESCRIPTION
**Pull request by sahilpalvia**
*Issue #, if available:*
#399 

*Description of changes:*
* Emitting logging messages at DEBUG level for retryable exceptions
* Introducing SubscribeToShardRetryableException

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java [hunk 1]
```diff
@@ -184,29 +198,55 @@ private void handleFlowError(Throwable t) {
...
+        ACUIRE_TIMEOUT("AcquireTimeout"), READ_TIMEOUT("ReadTimeout"), UNKNOWN("Unknown");
+
+        String value;
+
+        ThrowableType(final String value) {
+            this.value = value;
+        }
+    }
+
+    private class ThrowableCategory {
```
Comment (pfifer) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: You should make this class static as it doesn't need to close over `this`

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java [hunk 2]
```diff
@@ -184,29 +198,55 @@ private void handleFlowError(Throwable t) {
         }
     }
 
-    private String throwableCategory(Throwable t) {
+    private enum ThrowableType {
+        ACUIRE_TIMEOUT("AcquireTimeout"), READ_TIMEOUT("ReadTimeout"), UNKNOWN("Unknown");
```
Comment (pfifer) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Probably should be `OTHER` vs `UNKNOWN`.  Unknown feels like it's a failure state (subjective I know).

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java [hunk 3]
```diff
@@ -184,29 +198,55 @@ private void handleFlowError(Throwable t) {
...
+        }
+    }
+
+    private ThrowableCategory throwableCategory(Throwable t) {
         Throwable current = t;
         StringBuilder builder = new StringBuilder();
         do {
             if (current.getMessage() != null && current.getMessage().startsWith("Acquire operation")) {
-                return "AcquireTimeout";
+                return new ThrowableCategory(ThrowableType.ACUIRE_TIMEOUT);
```
Comment (pfifer) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: These should just be static instances of the ThrowableCategory as they have no local state.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/SubscribeToShardRetryableException.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package software.amazon.kinesis.retrieval.fanout;
+
+import software.amazon.kinesis.exceptions.KinesisClientLibRetryableException;
+
+/**
+ * RetryableException for SubscribeToShard APIs.
+ */
+public class SubscribeToShardRetryableException extends KinesisClientLibRetryableException {
```
Comment (pfifer) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/SubscribeToShardRetryableException.java]: This should probably be in `retrieval` and called `RetryableRetrievalException`.  The `ShardConsumer` shouldn't know that it's using SubscribeToShard.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java [hunk 5]
```diff
@@ -25,7 +25,10 @@
 import org.reactivestreams.Subscriber;
 import org.reactivestreams.Subscription;
 
+import lombok.Data;
+import lombok.NonNull;
 import lombok.RequiredArgsConstructor;
+import lombok.experimental.Accessors;
```
Comment (pfifer) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Looks like these imports were added, but not used.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java (hunk 1 above)
Comment (sahilpalvia) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Addressed.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java (hunk 2 above)
Comment (sahilpalvia) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Thought naming it other earlier, but ended up naming it unknown. Changed it to other.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java (hunk 3 above)
Comment (sahilpalvia) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Addressed

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/SubscribeToShardRetryableException.java (hunk 4 above)
Comment (sahilpalvia) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/SubscribeToShardRetryableException.java]: Addressed.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java (hunk 5 above)
Comment (sahilpalvia) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Not sure why organize imports was ignored.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java [hunk 6]
```diff
@@ -184,29 +200,55 @@ private void handleFlowError(Throwable t) {
         }
     }
 
-    private String throwableCategory(Throwable t) {
+    private enum ThrowableType {
+        ACUIRE_TIMEOUT("AcquireTimeout"), READ_TIMEOUT("ReadTimeout"), OTHER("Other");
```
Comment (emlyn) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Not that it matters too much, as it's private, but there's a typo here: `ACUIRE_TIMEOUT` -> `ACQUIRE_TIMEOUT` @sahilpalvia @pfifer

Comment (pfifer) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java]: Thanks for reporting this, we'll fix it in a later release.

**Pull request merged**